File: back/app/schemas/roles.py
"""
Схемы для работы с ролями
"""

from pydantic import BaseModel, field_validator
from typing import Optional, Dict, Any
from datetime import datetime
import json

# Системные роли (используем существующую UserRole)
class UserRoleBase(BaseModel):
    code: str
    name: str
    name_native: Optional[str] = None
    name_en: Optional[str] = None
    description: Optional[str] = None
    permissions: Optional[Dict[str, Any]] = None
    is_active: bool = True

class UserRoleCreate(UserRoleBase):
    pass

class UserRoleUpdate(BaseModel):
    name: Optional[str] = None
    name_native: Optional[str] = None
    name_en: Optional[str] = None
    description: Optional[str] = None
    permissions: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None

class UserRoleResponse(UserRoleBase):
    id: int
    created_at: datetime
    
    class Config:
        from_attributes = True

# Проектные роли
class ProjectRoleBase(BaseModel):
    code: str
    name: str
    name_en: Optional[str] = None
    description: Optional[str] = None
    permissions: Optional[Dict[str, Any]] = None
    is_active: bool = True

class ProjectRoleCreate(ProjectRoleBase):
    pass

class ProjectRoleUpdate(BaseModel):
    name: Optional[str] = None
    name_en: Optional[str] = None
    description: Optional[str] = None
    permissions: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None

class ProjectRoleResponse(ProjectRoleBase):
    id: int
    created_at: datetime
    
    class Config:
        from_attributes = True
    
    @field_validator('permissions', mode='before')
    @classmethod
    def parse_permissions(cls, v):
        """Parse JSON string permissions to dictionary"""
        if v is None:
            return None
        if isinstance(v, str):
            try:
                return json.loads(v)
            except (json.JSONDecodeError, TypeError):
                return None
        return v

# Схемы для пользователей с ролями
class UserWithUserRole(BaseModel):
    id: int
    username: str
    email: str
    full_name: str
    is_active: bool
    user_role: Optional[UserRoleResponse] = None
    
    class Config:
        from_attributes = True

# Схемы для участников проекта с ролями
class ProjectMemberWithRole(BaseModel):
    id: int
    project_id: int
    user_id: int
    role: str  # Legacy field
    project_role: Optional[ProjectRoleResponse] = None
    permissions: Optional[str] = None
    joined_at: Optional[datetime] = None
    user: Optional[UserWithUserRole] = None
    
    class Config:
        from_attributes = True

File: front/src/types/roles.ts
// Типы для работы с ролями

export interface SystemRole {
  id: number;
  code: string;
  name: string;
  name_en?: string;
  description?: string;
  permissions?: Record<string, any>;
  is_active: boolean;
  created_at: string;
}

export interface ProjectRole {
  id: number;
  code: string;
  name: string;
  name_en?: string;
  description?: string;
  permissions?: Record<string, any>;
  is_active: boolean;
  created_at: string;
}

export interface UserWithSystemRole {
  id: number;
  username: string;
  email: string;
  full_name: string;
  is_active: boolean;
  system_role?: SystemRole;
}

export interface ProjectMemberWithRole {
  id: number;
  project_id: number;
  user_id: number;
  role: string; // Legacy field
  project_role?: ProjectRole;
  permissions?: string;
  joined_at?: string;
  user?: UserWithSystemRole;
}

// Константы для системных ролей
export const SYSTEM_ROLES = {
  ADMIN: 'admin',
  OPERATOR: 'operator',
  VIEWER: 'viewer'
} as const;

export type SystemRoleCode = typeof SYSTEM_ROLES[keyof typeof SYSTEM_ROLES];

// Константы для проектных ролей
export const PROJECT_ROLES = {
  OWNER: 'owner',
  MANAGER: 'manager',
  REVIEWER: 'reviewer',
  CONTRIBUTOR: 'contributor',
  VIEWER: 'viewer'
} as const;

export type ProjectRoleCode = typeof PROJECT_ROLES[keyof typeof PROJECT_ROLES];

// Права доступа для системных ролей
export const SYSTEM_ROLE_PERMISSIONS = {
  [SYSTEM_ROLES.ADMIN]: {
    canCreateProjects: true,
    canEditProjects: true,
    canDeleteProjects: true,
    canViewProjects: true,
    canCreateDocuments: true,
    canEditDocuments: true,
    canDeleteDocuments: true,
    canViewDocuments: true,
    canCreateTransmittals: true,
    canEditTransmittals: true,
    canDeleteTransmittals: true,
    canViewTransmittals: true,
    canCreateReviews: true,
    canEditReviews: true,
    canDeleteReviews: true,
    canViewReviews: true,
    canManageUsers: true,
    canViewUsers: true,
    canViewDashboard: true,
    canManageSystemSettings: true
  },
  [SYSTEM_ROLES.OPERATOR]: {
    canCreateProjects: true,
    canEditProjects: true,
    canDeleteProjects: false,
    canViewProjects: true,
    canCreateDocuments: true,
    canEditDocuments: true,
    canDeleteDocuments: true,
    canViewDocuments: true,
    canCreateTransmittals: true,
    canEditTransmittals: true,
    canDeleteTransmittals: false,
    canViewTransmittals: true,
    canCreateReviews: true,
    canEditReviews: true,
    canDeleteReviews: false,
    canViewReviews: true,
    canManageUsers: false,
    canViewUsers: true,
    canViewDashboard: true,
    canManageSystemSettings: false
  },
  [SYSTEM_ROLES.VIEWER]: {
    canCreateProjects: false,
    canEditProjects: false,
    canDeleteProjects: false,
    canViewProjects: true,
    canCreateDocuments: false,
    canEditDocuments: false,
    canDeleteDocuments: false,
    canViewDocuments: true,
    canCreateTransmittals: false,
    canEditTransmittals: false,
    canDeleteTransmittals: false,
    canViewTransmittals: true,
    canCreateReviews: false,
    canEditReviews: false,
    canDeleteReviews: false,
    canViewReviews: true,
    canManageUsers: false,
    canViewUsers: false,
    canViewDashboard: true,
    canManageSystemSettings: false
  }
} as const;

// Права доступа для проектных ролей
export const PROJECT_ROLE_PERMISSIONS = {
  [PROJECT_ROLES.OWNER]: {
    canManageProject: true,
    canManageMembers: true,
    canCreateDocuments: true,
    canEditDocuments: true,
    canDeleteDocuments: true,
    canViewDocuments: true,
    canCreateTransmittals: true,
    canEditTransmittals: true,
    canDeleteTransmittals: true,
    canViewTransmittals: true,
    canCreateReviews: true,
    canEditReviews: true,
    canDeleteReviews: true,
    canViewReviews: true,
    canApproveDocuments: true,
    canRejectDocuments: true
  },
  [PROJECT_ROLES.MANAGER]: {
    canManageProject: false,
    canManageMembers: false,
    canCreateDocuments: true,
    canEditDocuments: true,
    canDeleteDocuments: true,
    canViewDocuments: true,
    canCreateTransmittals: true,
    canEditTransmittals: true,
    canDeleteTransmittals: true,
    canViewTransmittals: true,
    canCreateReviews: true,
    canEditReviews: true,
    canDeleteReviews: true,
    canViewReviews: true,
    canApproveDocuments: true,
    canRejectDocuments: true
  },
  [PROJECT_ROLES.REVIEWER]: {
    canManageProject: false,
    canManageMembers: false,
    canCreateDocuments: false,
    canEditDocuments: false,
    canDeleteDocuments: false,
    canViewDocuments: true,
    canCreateTransmittals: false,
    canEditTransmittals: false,
    canDeleteTransmittals: false,
    canViewTransmittals: true,
    canCreateReviews: true,
    canEditReviews: true,
    canDeleteReviews: false,
    canViewReviews: true,
    canApproveDocuments: true,
    canRejectDocuments: true
  },
  [PROJECT_ROLES.CONTRIBUTOR]: {
    canManageProject: false,
    canManageMembers: false,
    canCreateDocuments: true,
    canEditDocuments: true,
    canDeleteDocuments: false,
    canViewDocuments: true,
    canCreateTransmittals: true,
    canEditTransmittals: true,
    canDeleteTransmittals: false,
    canViewTransmittals: true,
    canCreateReviews: false,
    canEditReviews: false,
    canDeleteReviews: false,
    canViewReviews: true,
    canApproveDocuments: false,
    canRejectDocuments: false
  },
  [PROJECT_ROLES.VIEWER]: {
    canManageProject: false,
    canManageMembers: false,
    canCreateDocuments: false,
    canEditDocuments: false,
    canDeleteDocuments: false,
    canViewDocuments: true,
    canCreateTransmittals: false,
    canEditTransmittals: false,
    canDeleteTransmittals: false,
    canViewTransmittals: true,
    canCreateReviews: false,
    canEditReviews: false,
    canDeleteReviews: false,
    canViewReviews: true,
    canApproveDocuments: false,
    canRejectDocuments: false
  }
} as const;

// Функции для проверки прав
export const hasSystemPermission = (
  userRole: SystemRoleCode, 
  permission: keyof typeof SYSTEM_ROLE_PERMISSIONS[SystemRoleCode]
): boolean => {
  return SYSTEM_ROLE_PERMISSIONS[userRole]?.[permission] ?? false;
};

export const hasProjectPermission = (
  projectRole: ProjectRoleCode, 
  permission: keyof typeof PROJECT_ROLE_PERMISSIONS[ProjectRoleCode]
): boolean => {
  return PROJECT_ROLE_PERMISSIONS[projectRole]?.[permission] ?? false;
};
